futures-async-runtime/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+#![cfg_attr(feature = "nightly", feature(conservative_impl_trait))]
+#![cfg_attr(feature = "nightly", feature(generator_trait))]
+#![cfg_attr(feature = "nightly", feature(on_unimplemented))]
+#![cfg_attr(feature = "nightly", feature(arbitrary_self_types))]
+#![cfg_attr(feature = "nightly", feature(optin_builtin_traits))]
+#![cfg_attr(feature = "nightly", feature(pin))]
+
+macro_rules! if_nightly {
+    ($($i:item)*) => ($(
+        #[cfg(feature = "nightly")]
+        $i
+    )*)
+}
+
+if_nightly! {
+    extern crate futures_core;
+    extern crate futures_stable;
+
+    mod future;
+    mod stream;
+    mod pinned_future;
+    mod pinned_stream;
+
+    use std::cell::Cell;
+    use std::mem;
+    use std::ptr;
+    use futures_core::task;
+
+    pub use self::future::*;
+    pub use self::stream::*;
+    pub use self::pinned_future::*;
+    pub use self::pinned_stream::*;
+
+    pub use futures_core::{Async, Future, Stream};
+    pub use futures_stable::{StableFuture, StableStream};
+
+    pub use std::ops::Generator;
+
+#[rustc_on_unimplemented = "async functions must return a `Result` or \
+                                a typedef of `Result`"]
+    pub trait IsResult {
```

Comment:
Out of curiosity, is the plan to move this trait into the standard library at one point? It looks like it could serve as the base trait for `Try`, so that:

```rust
pub trait Try: IsResult {
    fn from_ok(v: Self::Ok) -> Self;
    fn from_err(v: Self::Err) -> Self;
}
```

Reply:
This is just a hack to make error messages better, its not supposed to be a part of the API for end users.

Reply:
I was actually recently wondering if this could be extended to `Try` instead of forcing `Result`. Currently that would only additionally allow creating an `impl Future<Item=T, Error=NoneError>` from an `Option<T>` from `std`, but I wonder if there are any interesting implementations of `Try` outside the standard library?